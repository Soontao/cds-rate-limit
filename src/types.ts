import type { IRateLimiterOptions } from "rate-limiter-flexible";

export type KeyPart = "user_id" | "remote_ip" | "tenant";

export type AnonymousPolicy = false | Pick<RateLimitOptions, "duration" | "points" | "keyPrefix">

export interface RateLimitOptions extends IRateLimiterOptions {
  impl?: "memory" | "redis";
  keyParts?: Array<KeyPart>;
  /**
   * policy for anonymous requests
   */
  anonymous?: AnonymousPolicy;
}

export interface MemoryRateLimitOptions extends RateLimitOptions {
  impl: "memory";
}

export interface RedisRateLimitOptions extends RateLimitOptions {
  impl: "redis";
  /**
   * redis client
   */
  storeClient: any;
}
