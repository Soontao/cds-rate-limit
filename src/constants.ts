
export const ANNOTATE_CDS_RATE_LIMIT = "@cds.rate.limit";

export const FLAG_RATE_LIMIT_PERFORMED = "__rate_limit_performed";

export const GLOBAL_RATE_LIMITER_PREFIX = "global-rate-limiter";

export const RATE_LIMIT_HEADERS = {
  "Retry-After": "Retry-After",
  "X-RateLimit-Limit": "X-RateLimit-Limit",
  "X-RateLimit-Remaining": "X-RateLimit-Remaining",
  "X-RateLimit-Reset": "X-RateLimit-Reset",
} as const;
