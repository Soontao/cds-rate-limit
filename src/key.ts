import { KeyPart } from "./types";



/**
 * custom key extractor
 */
export interface KeyExtractor {
  /**
   * extract key for target key part label
   */
  targetKeyPart: string;
  /**
   * extract key part from cds event/context
   */
  extract: (evt: any) => string;
}

export const builtInExtractors: Array<KeyExtractor> = [
  {
    targetKeyPart: "tenant",
    extract: (evt) => evt?.tenant ?? "unknown_tenant"
  },
  {
    targetKeyPart: "user_id",
    extract: (evt) => evt?.user?.id ?? "unknown_user"
  },
  {
    targetKeyPart: "remote_ip",
    extract: (evt) => evt?._?.req?.ip ?? evt?._?.req?.socket.remoteAddress ?? "unknown_ip"
  },
];

export const keyExtractorCreatorBuilder = (extensionExtractors: Array<KeyExtractor> = []) => (keyParts: Array<KeyPart>) => (evt: any) => {
  const parts = [];

  for (const extractor of [...builtInExtractors, ...extensionExtractors]) {
    if (extractor.targetKeyPart !== undefined && keyParts.includes(extractor.targetKeyPart)) {
      parts.push(extractor.extract(evt));
    }
  }

  return parts.join("/");
};
