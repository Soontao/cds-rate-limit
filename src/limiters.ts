import { RateLimiterAbstract, RateLimiterMemory, RateLimiterRedis } from "rate-limiter-flexible";
import { RateLimitOptions } from "./types";

const limiters = new Map();

export const provisionRateLimiter = (options: RateLimitOptions): RateLimiterAbstract => {
  if (!limiters.has(options.keyPrefix)) {
    switch (options.impl) {
      case "redis":
        // fallback to in memory rate limiter when redis is not available
        limiters.set(
          options.keyPrefix,
          new RateLimiterRedis(Object.assign({}, options as any, { insuranceLimiter: new RateLimiterMemory(options) }))
        );
      default:
        limiters.set(
          options.keyPrefix,
          new RateLimiterMemory(options)
        );
    }
  }
  return limiters.get(options.keyPrefix);
};
