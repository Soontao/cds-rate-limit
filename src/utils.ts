
export const groupByKeyPrefix = (obj: any, key: string) => {
  if (obj === undefined || obj === null) {
    return {};
  }
  const keys = Object.keys(obj);
  return keys
    .filter(objectKey => objectKey.startsWith(key))
    .reduce((pre: any, cur: string) => {
      pre[cur.substring(key.length + 1)] = obj[cur];
      return pre;
    }, {});

};
