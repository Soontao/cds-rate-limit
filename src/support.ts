import { LRUCacheProvider } from "@newdash/newdash/cacheProvider";
import { IRateLimiterOptions, RateLimiterAbstract, RateLimiterMemory, RateLimiterRes } from "rate-limiter-flexible";
import { FLAG_RATE_LIMIT_PERFORMED, GLOBAL_RATE_LIMITER_PREFIX, RATE_LIMIT_HEADERS } from "./constants";


export type KeyPart = "user_id" | "remote_ip" | "tenant";

export interface RateLimitOptions extends IRateLimiterOptions {
  impl?: "memory" | "redis";
  keyParts?: Array<KeyPart>;
}

export const DEFAULT_OPTIONS: RateLimitOptions = {
  impl: "memory",
  keyParts: ["tenant"],
  keyPrefix: GLOBAL_RATE_LIMITER_PREFIX,
  duration: 60, // 60 seconds
  points: 200 * 60, // 200 requests per seconds
};

const createKeyExtractor = (keyParts: Array<KeyPart>) => (evt: any) => {
  const parts = [];
  if (keyParts.includes("tenant")) {
    parts.push(evt?.tenant ?? "unknown_tenant");
  }
  if (keyParts.includes("user_id")) {
    parts.push(evt?.user?.id ?? "unknown_user");
  }
  if (keyParts.includes("remote_ip")) {
    const req = evt?._?.req;
    if (req !== undefined) {
      parts.push(req?.ip ?? req.socket.remoteAddress ?? "unknown_ip");
    }
  }
  return parts.join("/");
};

/**
 * parse options for events
 * 
 * @param evt 
 * @param globalOptions 
 * @returns 
 */
const parseOptions = (evt: any, globalOptions: RateLimitOptions): RateLimitOptions => {
  // TODO: if not have configured, return default
  return Object.assign({}, globalOptions);
};

/**
 * get the rate limiter key from event 
 * 
 * @param service 
 * @param evt 
 * @returns 
 */
const extractRateLimiterKey = (service: any, evt: any): string => {
  const rateLimiterKey = `${service.name}/${evt?.entity ?? "unbound"}/${evt}`;
  // TODO: if not have configured, return GLOBAL
  return rateLimiterKey;
};

const provisionRateLimiter = (options: RateLimitOptions): RateLimiterAbstract => {
  switch (options.impl) {
    default:
      return new RateLimiterMemory(options);
  }
};

/**
 * attach headers to evt
 * 
 * @param evt 
 * @param total 
 * @param rateLimitRes 
 */
const attachHeaders = (evt: any, total: number, rateLimitRes: RateLimiterRes) => {
  const response = evt?._?.req?.res;
  if (response !== undefined) {
    response?.set?.({
      [RATE_LIMIT_HEADERS["Retry-After"]]: Math.floor(rateLimitRes.msBeforeNext / 1000),
      [RATE_LIMIT_HEADERS["X-RateLimit-Limit"]]: total,
      [RATE_LIMIT_HEADERS["X-RateLimit-Remaining"]]: rateLimitRes.remainingPoints,
      [RATE_LIMIT_HEADERS["X-RateLimit-Reset"]]: Math.floor((Date.now() + rateLimitRes.msBeforeNext) / 1000)
    });
  }
};

/**
 * apply rate limitation for cds
 * 
 * @param cds 
 * @param globalOptions 
 */
export const applyRateLimit = (cds: any, globalOptions: RateLimitOptions = {}) => {
  /**
   * key: service/event id
   */
  const limiters = new LRUCacheProvider(10240);

  globalOptions = Object.assign({}, DEFAULT_OPTIONS, globalOptions ?? {});

  cds.on("serving", (service: any) => {

    if (service instanceof cds.ApplicationService) { // only application services

      const logger = cds.log(service?.name);

      service.before("*", async (evt: any) => {

        const rateLimiterKey = extractRateLimiterKey(service, evt);
        // only affect HTTP requests
        if (evt instanceof cds.Request) {
          // if this event has been measured
          if (evt[FLAG_RATE_LIMIT_PERFORMED] === true) {
            logger.debug(
              "event",
              rateLimiterKey,
              "is triggered by internal communication (has been measured by first event), ignored"
            );
            return;
          }

          evt[FLAG_RATE_LIMIT_PERFORMED] = true;

          const options = parseOptions(evt, globalOptions);
          const rateLimiter = limiters.getOrCreate(rateLimiterKey, () => provisionRateLimiter(options));
          const keyExtractor = createKeyExtractor(options.keyParts as []);
          const key = keyExtractor(evt);

          try {
            const response = await rateLimiter.consume(key);
            logger.debug("rate limit consume successful:", key, response);
            attachHeaders(evt, options.points as number, response);
            return;
          } catch (response) {
            logger.error("rate limit consume failed:", key, response);
            attachHeaders(evt, options.points as number, response);
            return evt.reject(
              429,
              `Rate limit exceed, please retry after ${Math.floor(response.msBeforeNext / 1000)} seconds`
            );
          }
        } else {
          logger.debug("event", rateLimiterKey, "is not from http request, ignored");
        }
      });

    }
  });

};
