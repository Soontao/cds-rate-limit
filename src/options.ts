import { isEmpty } from "@newdash/newdash/isEmpty";
import { ANNOTATE_CDS_RATE_LIMIT } from "./constants";
import { RateLimitOptions } from "./types";
import { cwdRequire, formatEventKey, groupByKeyPrefix } from "./utils";

const optionsCache = new Map<string, RateLimitOptions>();

/**
 * parse options for events
 * 
 * @param evt 
 * @param globalOptions 
 * @returns 
 */
export const parseOptions = (service: any, evt: any, globalOptions: RateLimitOptions): RateLimitOptions => {

  const key = formatEventKey(service, evt);
  if (!optionsCache.has(key)) {
    const cds = cwdRequire("@sap/cds");

    const localOptions: RateLimitOptions = {};

    const serviceLevelOptions = groupByKeyPrefix(service.definition, ANNOTATE_CDS_RATE_LIMIT);

    if (!isEmpty(serviceLevelOptions)) {
      if ("duration" in serviceLevelOptions || "points" in serviceLevelOptions) {
        // use service key-prefix, which will be used to indicate the Rate Limiter
        localOptions.keyPrefix = `local-${service.name}`;
      }
      Object.assign(localOptions, serviceLevelOptions);
    }

    /**
     * definition for action/function
     */
    let actionFunctionDef = undefined;

    // entity relevant
    if (evt?.entity !== undefined) {
      // entity related events, CRUD
      const entityDef = cds.model.definitions[evt.entity];
      const entityLevelOptions = groupByKeyPrefix(entityDef, ANNOTATE_CDS_RATE_LIMIT);
      if (!isEmpty(entityLevelOptions)) {
        if ("duration" in entityLevelOptions || "points" in entityLevelOptions) {
          // use entity key-prefix, which will be used to indicate the Rate Limiter
          localOptions.keyPrefix = `local-${service.name}/${evt?.entity}`;
        }
        Object.assign(localOptions, entityLevelOptions);
      };

      // bound action/function
      if (entityDef.actions !== undefined && evt.event in entityDef.actions) {
        actionFunctionDef = entityDef.actions[evt.event];
      }
    }

    // unbound action/function
    if (evt.entity === undefined && evt.event in service.operations()) {
      actionFunctionDef = service.operations()[evt.event];
    }

    if (actionFunctionDef !== undefined) {
      const eventLevelOptions = groupByKeyPrefix(actionFunctionDef, ANNOTATE_CDS_RATE_LIMIT);
      if (!isEmpty(eventLevelOptions)) {
        if ("duration" in eventLevelOptions || "points" in eventLevelOptions) {
          // use entity key-prefix, which will be used to indicate the Rate Limiter
          localOptions.keyPrefix = `local-${service.name}/${evt?.entity}/${evt.event}`;
        }
        Object.assign(localOptions, eventLevelOptions);
      };
    }

    optionsCache.set(key, Object.assign({}, globalOptions, localOptions));
  }

  return optionsCache.get(key) as any;

};