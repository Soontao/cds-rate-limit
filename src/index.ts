import { cwdRequireCDS } from "cds-internal-tool";
import { applyRateLimit } from "./support";

applyRateLimit(cwdRequireCDS());
