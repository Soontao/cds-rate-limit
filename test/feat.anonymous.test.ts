
import { setupTest } from "cds-internal-tool";
import { RATE_LIMIT_HEADERS } from "../src/constants";

describe("Support Anonymous User Test Suite", () => {

  const axios = setupTest(__dirname, "./app")

  it('should support RateLimit headers', async () => {
    let responses = await axios.get("/sample/People")
    expect(responses.status).toBe(401)
    expect(responses.headers[RATE_LIMIT_HEADERS["Retry-After"].toLowerCase()]).toMatchSnapshot()
    expect(responses.headers[RATE_LIMIT_HEADERS["X-RateLimit-Limit"].toLowerCase()]).toMatchSnapshot()
    expect(responses.headers[RATE_LIMIT_HEADERS["X-RateLimit-Remaining"].toLowerCase()]).toMatchSnapshot()
    expect(responses.headers[RATE_LIMIT_HEADERS["X-RateLimit-Reset"].toLowerCase()]).not.toBeUndefined()

    // for another ip
    responses = await axios.get("/sample/People", {
      headers: {
        'X-Forwarded-For': '1.1.1.1'
      }
    })

    expect(responses.status).toBe(401)
    expect(responses.headers[RATE_LIMIT_HEADERS["Retry-After"].toLowerCase()]).toMatchSnapshot()
    expect(responses.headers[RATE_LIMIT_HEADERS["X-RateLimit-Limit"].toLowerCase()]).toMatchSnapshot()
    expect(responses.headers[RATE_LIMIT_HEADERS["X-RateLimit-Remaining"].toLowerCase()]).toMatchSnapshot()
    expect(responses.headers[RATE_LIMIT_HEADERS["X-RateLimit-Reset"].toLowerCase()]).not.toBeUndefined()

  });

});
