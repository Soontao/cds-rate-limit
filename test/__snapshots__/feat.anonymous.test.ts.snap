// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Support Anonymous User Test Suite should support RateLimit headers 1`] = `"60"`;

exports[`Support Anonymous User Test Suite should support RateLimit headers 2`] = `"100"`;

exports[`Support Anonymous User Test Suite should support RateLimit headers 3`] = `"99"`;

exports[`Support Anonymous User Test Suite should support RateLimit headers 4`] = `"60"`;

exports[`Support Anonymous User Test Suite should support RateLimit headers 5`] = `"100"`;

exports[`Support Anonymous User Test Suite should support RateLimit headers 6`] = `"99"`;
